File: src/main/java/io/disassemble/javanalysis/CtMethodNode.java
package io.disassemble.javanalysis;

import io.disassemble.javanalysis.flow.ControlFlowGraph;
import io.disassemble.javanalysis.flow.FlowBlock;
import io.disassemble.javanalysis.insn.CtInsn;
import io.disassemble.javanalysis.util.CodeParser;
import javassist.CtMethod;
import javassist.bytecode.CodeAttribute;
import javassist.bytecode.ConstPool;
import javassist.bytecode.MethodInfo;
import javassist.bytecode.analysis.ControlFlow;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author Tyler Sedlar
 * @since 5/19/2017
 */
public class CtMethodNode {

    private final CtMethod source;
    private final CtClassNode owner;
    private List<CtInsn> instructions;

    private final Map<Integer, Integer> indices = new HashMap<>();

    public CtMethodNode(CtMethod source, CtClassNode owner) {
        this.source = source;
        this.owner = owner;
    }

    public CtMethod source() {
        return source;
    }

    public CtClassNode owner() {
        return owner;
    }

    public String name() {
        return source.getName();
    }

    public String desc() {
        return source.getSignature();
    }

    public String key() {
        return owner.name() + "." + name() + desc();
    }

    public MethodInfo info() {
        return source.getMethodInfo();
    }

    public CodeAttribute code() {
        return info().getCodeAttribute();
    }

    public ConstPool pool() {
        return info().getConstPool();
    }

    public List<CtInsn> instructions() {
        if (instructions != null) {
            return instructions;
        }
        return (instructions = CodeParser.parse(this));
    }

    public int line() {
        return instructions().get(0).line() - 1;
    }

    public void index(int insnIndex, int position) {
        indices.put(insnIndex, position);
    }

    public int indexOf(CtInsn insn) {
        return normalizeIndex(insn.index());
    }

    public int normalizeIndex(int index) {
        return (indices.containsKey(index) ? indices.get(index) : indices.get(index - 1));
    }

    public boolean hasIndex(int index) {
        return indices.containsKey(index);
    }

    public ControlFlowGraph cfg() {
        try {
            ControlFlow flow = new ControlFlow(owner.source(), info());
            List<FlowBlock> blocks = new ArrayList<>();
            for (ControlFlow.Block bb : flow.basicBlocks()) {
                blocks.add(new FlowBlock(this, bb));
            }
            ControlFlowGraph cfg = new ControlFlowGraph(flow, blocks);
            blocks.forEach(block -> block.init(cfg));
            return cfg;
        } catch (Exception e) {
            e.printStackTrace();
            throw new IllegalStateException("Invalid bytecode for cfg in " + source.getLongName());
        }
    }
}

File: src/main/java/io/disassemble/javanalysis/CtMethodNode.kt
package io.disassemble.javanalysis

import io.disassemble.javanalysis.flow.ControlFlowGraph
import io.disassemble.javanalysis.flow.FlowBlock
import io.disassemble.javanalysis.insn.CtInsn
import io.disassemble.javanalysis.util.CodeParser
import javassist.CtMethod
import javassist.bytecode.CodeAttribute
import javassist.bytecode.ConstPool
import javassist.bytecode.MethodInfo
import javassist.bytecode.analysis.ControlFlow
import java.util.*

/**
 * @author Tyler Sedlar
 * @since 5/19/2017
 */
class CtMethodNode(private val source: CtMethod, private val owner: CtClassNode) {
    private val instructions: List<CtInsn> by lazy { CodeParser.parse(this) }

    private val indices = HashMap<Int, Int>()

    fun source(): CtMethod {
        return source
    }

    fun owner(): CtClassNode {
        return owner
    }

    fun name(): String {
        return source.name
    }

    fun desc(): String {
        return source.signature
    }

    fun key(): String {
        return "${owner.name()}.${name()}${desc()}"
    }

    fun info(): MethodInfo {
        return source.methodInfo
    }

    fun code(): CodeAttribute {
        return info().codeAttribute
    }

    fun pool(): ConstPool {
        return info().constPool
    }

    fun instructions(): List<CtInsn> {
        return instructions
    }

    fun line(): Int {
        return instructions()[0].line() - 1
    }

    fun index(insnIndex: Int, position: Int) {
        indices[insnIndex] = position
    }

    fun indexOf(insn: CtInsn): Int {
        return normalizeIndex(insn.index())
    }

    fun normalizeIndex(index: Int): Int {
        return if (index in indices) {
            indices[index]
        } else {
            indices[index - 1]
        } ?: error("Unable to normalize index: $index")
    }

    fun hasIndex(index: Int): Boolean {
        return indices.containsKey(index)
    }

    fun cfg(): ControlFlowGraph {
        try {
            val flow = ControlFlow(owner.source(), info())
            val blocks = flow.basicBlocks().map { FlowBlock(this, it) }
            val cfg = ControlFlowGraph(flow, blocks)
            blocks.forEach { block -> block.init(cfg) }
            return cfg
        } catch (e: Exception) {
            e.printStackTrace()
            throw IllegalStateException("Invalid bytecode for cfg in " + source.longName)
        }

    }
}
